An attempt to remove traits

mateuszbaran opened the pull request:
The goal is to make the internal structure easier to understand.

kellertuer committed:
Setup runic.

mateuszbaran committed:
Trait removal pt.1: IsQuotientManifold

kellertuer committed:
Removing Traits Part II: IsDefaultMetric -> default_metric(M)

mateuszbaran committed:
Removing IsEmbeddedManifold

mateuszbaran committed:
Merge remote-tracking branch 'origin/mbaran/trait-reduction' into mbaran/trait-reduction

kellertuer committed:
Some further to removing the `DefaultMetric` trait.

mateuszbaran commented:
It's (hopefully) small from the point of view of end users. Feel free to remove it if you think it doesn't fit.

mateuszbaran committed:
Merge remote-tracking branch 'origin/mbaran/trait-reduction' into mbaran/trait-reduction

kellertuer committed:
Fix some more metric functions.

mateuszbaran committed:
Merge remote-tracking branch 'origin/mbaran/trait-reduction' into mbaran/trait-reduction

kellertuer committed:
A few further fixes.

kellertuer commented:
From that perspective yes, from ours, no ;) Internally it is quite a rework, but I like that the `default_` function approach gets more unified now as well.

kellertuer committed:
Fix basis and exp_fused!-passdown-default

kellertuer commented:
I fixed all manifolds generally,
in `tests/metric.jl` there is still a lot of failings, since the interaction with the `IsMetricManifold` has to be reworked. Some of that is the default to solve an ODE for metric manifolds. 

I have to think a bit how to rewrite that logic so that the function `exp!` in line 52 `ManifoldsOrdinaryDiffEqExt.jl` is accordingly “activated” once one loads that package. Currently still the new version without traits that checks for default metric is called even after extension is loaded.
I would expect that the interaction / removing `IsMetricManifold` would solve most those remaining errors in metric.jl tests.

kellertuer committed:
Remove the IsMetricManifold trait.

kellertuer committed:
Start a News entry.

kellertuer committed:
Fix a few traits functions to return lists.

kellertuer committed:
Fix two typos that affected the docs.

kellertuer commented:
I am running into a small problem with the “dispatch logic” for `exp` on metric manifolds without traits.

Formerly:
* the metric that is now the `default_metric` would pass on to the underlying manifold (with implicit metric)
* all others would error usually,
* some “magically” called the ODE solver when `IsMetricManifold` was set, though that is a retraction actually

The last two can not really be distinguished by now. I would propose to leave out the last one as a “magic default” since if no exp is implemented, the user has to specify a retraction anyways – but could use the ODEsolver one?

kellertuer committed:
Fix a few small things, add a comment.

kellertuer committed:
remove more traits code

especially the recursion breaking is no longer necessary.

kellertuer commented:
I continued removing the Connection based traits `IsDefaultConnection` and `IsConnectionManifold`. I noticed that there is a `connection(M)` function, so I am now considering calling `default_metric(M)` also just `metric(M)` since that seems not only more consistent but also more in line with: For every manifold we assume a metric and only one, so there is no need for a `metric_` case.

I also fixed the default of the ODE solver to use the `ODEExponentialRetraction` instead of overwriting `exp_fused!` and also defined that for both retract and its fused version – I think we might have missed to define the first one when we split those two functions.

There is still a few errors in the tests, but it is getting better and I think also overall more consistent.

kellertuer committed:
rename `default_metric` to `metric`, remove definitions of active traits.

mateuszbaran committed:
some work on embeddings

mateuszbaran committed:
Merge remote-tracking branch 'origin/mbaran/trait-reduction' into mbaran/trait-reduction

mateuszbaran committed:
formatting

kellertuer commented:
> I'm fine with removing groups as soon as https://github.com/JuliaManifolds/LieGroups.jl/issues/52 is resolved, but not earlier.

Hm, ok. That issue might still take a while, because it is a lot of types that I never felt had any use actually. But sure we should also check that for each case.

But who knows how long this rework will take, maybe I also find time until then to check at least a few functions. Or we do two breaking changes here in the end.

mateuszbaran commented:
I think I'd prefer to align removal of traits with splitting out Lie groups. BTW, not all things on that list are actions. There are also Cartan-Schouten connections or some Jacobians.

kellertuer commented:
Sure. I also specifically only referred to actions. For the connections I have to check, those I probably really simply forgot.

kellertuer committed:
remove a few active traits.

kellertuer committed:
Merge branch 'mbaran/trait-reduction' of github.com:JuliaManifolds/Manifolds.jl into mbaran/trait-reduction

# Conflicts:
#	src/manifolds/Elliptope.jl
#	src/manifolds/FixedRankMatrices.jl

kellertuer commented on src/groups/heisenberg.jl:
```diff
@@ -26,10 +26,11 @@ function HeisenbergGroup(n::Int; parameter::Symbol=:type)
     return HeisenbergGroup{typeof(size)}(size)
 end
 
+is_embedded_manifold(::HeisenbergGroup) = true
```
Just as a general remark, this function might need a case where it depends on the point type? Some representations might be embedded, others not. 

Though that might not affect this specific case, we at least should have that in mind and handle it similar to other cases like rand.

mateuszbaran replied:
That's a tricky point. It might indeed have to depend on the point type. I will think about it.

kellertuer commented:
> To be honest, I think that ODE solver for the geodesic is really bad. The one from `ManifoldsOrdinaryDiffEqDiffEqCallbacksExt` is the right approach. I'd prefer to just remove the old one. I'm thus perfectly fine with no longer using it automatically.

Hm I am by now very confused. But maybe that will just become my default status also on these PRs.
We do have a `solve_exp_ode!` in `ManifoldsOrdinaryDiffEqExt.jl` – with which I just fought for a while, bringing it into a retraction, since the `ODEExponentialRetraction` always requires a retraction internally to work, which does not fit to these. Besides that this looks totally fine to me? Why is it really bad? We even wrote that ODE in the Kalman paper. What is bad with that ODE?
I am tempted to introduce a retraction without an internal one for this ODE, but you said we can even remove it? I am confused, why we should remove the default ODE one solves for geodesics.

On the other hand `ManifoldsOrdinaryDiffEqDiffEqCallbacksExt.jl` works in charts? That I would consider the strange one then? Why is that the right approach?

Sorry, got lost on that and hence we have another day without much progress from my side. I struggle a bit coming from the code back to any theory, simply because there is none documented.

kellertuer committed:
Fix all but 2 metric dispatch logic things; those two might be related to embeddings actually.

kellertuer committed:
Now it is only one failing test where local metric is not found.

mateuszbaran commented:
> Besides that this looks totally fine to me? Why is it really bad? We even wrote that ODE in the Kalman paper. What is bad with that ODE?

The main issue is that it doesn't let the user select a chart to work in. It doesn't do proper chart switching. It basically assumes working in `Euclidean` with a custom metric. It uses `christoffel_symbols_second` which is an inefficient approach and only has an implicitly selected chart, in a way that has not been sufficiently verified to work well. And there is no point in verifying it because a better approach is already implemented.

The Kalman paper ODE is what's implemented in `ManifoldsOrdinaryDiffEqDiffEqCallbacksExt.jl`.

> I am tempted to introduce a retraction without an internal one for this ODE, but you said we can even remove it? I am confused, why we should remove the default ODE one solves for geodesics.

Because it's a poor default.

> On the other hand `ManifoldsOrdinaryDiffEqDiffEqCallbacksExt.jl` works in charts? That I would consider the strange one then? Why is that the right approach?

Yes, the geodesic equations are basically always expressed in a chart. What's so strange about it?

mateuszbaran committed:
a few improvements

mateuszbaran committed:
Merge remote-tracking branch 'origin/mbaran/trait-reduction' into mbaran/trait-reduction

mateuszbaran committed:
docs for the new geodesic solver

mateuszbaran committed:
add embedding type to SkewHermitianMatrices

mateuszbaran committed:
introduce embedding types

kellertuer committed:
This is not the right way to get the functions to hit the embedding – which one is?

tried for an hour with zero progess...

kellertuer committed:
Update src/manifolds/CenteredMatrices.jl

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

kellertuer committed:
Mark a test as no longer broken.

kellertuer committed:
Update test/differentiation.jl

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

mateuszbaran committed:
remove IsIsometricEmbeddedManifold, IsEmbeddedSubmanifold

mateuszbaran committed:
Merge remote-tracking branch 'origin/mbaran/trait-reduction' into mbaran/trait-reduction

kellertuer committed:
Fix skewhermitian, where two tests are now more precise.

kellertuer committed:
Improve one test.

kellertuer committed:
Mark a test as broken again that yesterday worked for short but now NaNs again.

kellertuer committed:
Update test/differentiation.jl

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

kellertuer committed:
Steps towards Grassmann

it seems exp “unwraps” Projector which it should not. It should not even forward that.

kellertuer committed:
In Theory switch to the right approx in Grassmann – in practice a few more things now fail ue to tolerances.

kellertuer committed:
raise default atol for Grassmann/StiefelPoints.

kellertuer committed:
Fix Grassmann.

kellertuer committed:
remove duplicate code.

kellertuer committed:
Fix determinant one.

kellertuer committed:
a small fix.

kellertuer committed:
Update src/manifolds/GeneralizedStiefel.jl

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

mateuszbaran committed:
fix FixedRank

mateuszbaran committed:
fix hyperbolic

mateuszbaran committed:
two more fixes

mateuszbaran committed:
fix ProjectiveSpace, remove unnecessary line

mateuszbaran committed:
Kendall's shape space doesn't currently have injectivity_radius implemented

mateuszbaran committed:
revert wrong change

kellertuer committed:
Fix multinomial SPD.

mateuszbaran commented:
I think I've figured out an even easier fix. I will push a commit to ManifoldsBase in a moment.

kellertuer committed:
Let MetricManifold forward embed/embed!, rand/rand! fixes all tests in Segre.

kellertuer committed:
Fix embedding type for sphere symmetric, they inherit even geodesics.

kellertuer committed:
It seems formerly `is_flat` was even transparent with respect to metric (=passthrough); that does not seem right. Fixed it for Stiefel.

kellertuer committed:
Fixed an accidental double doc string for FixedRank

kellertuer committed:
Step 1 in Fixing SPDs, some forwards are still a bit strange.

kellertuer commented:
Short status update: I think I fixed a few more manifolds (Segre, Stiefel), but now to a bit suck on the interplay between metric manifold and the embedding, or to be precise:

I think on the metric manifold, so using a nondefault metric, the non_simple embedding does not work correctly – or I missed to specify for this combination the unwrapping of SPDPoints (towards the embedding?) wrongly.

kellertuer committed:
I think I understood the new parameter. SPDs work.

kellertuer committed:
on sympossemidef the injectivity radius was tested - but that does not get inherited for isometric embeddings.

kellertuer committed:
Projector points on symplectic Grassmann require embedding.

kellertuer committed:
embed quaternions

kellertuer commented:
Short status update:

* `UnitaryMatrices` has a part on SpecialUmitary failing, (`is_flat`2x, `manifold_dimension`, and `)
* statistics has still failing parts that I also struggle to hunt down
* metric has still failing parts with the same problem as the previous two for me.

overall not many left (4 failed 5 errored)

I am unsure what to do about the GroupManifolds?

mateuszbaran committed:
fix allocation of zero_vector for Hyperbolic Poincare Ball manifold

kellertuer committed:
export one function less.

mateuszbaran committed:
remove groups and adapt to recent changes in ManifoldsBase.jl

kellertuer commented:
For Grassmann it is implicit enough that it can stay here.

Sure we can also move Kendall if you think we can not do it Grassmann-styleish enough.

kellertuer commented:
I would prefer keeping a Grassmann here, moving that would be super breaking, because it is used by quite a few people.

Sure if we can “enhance” Grassmann within LieGroups, that's cool, but I would do that as an enhancement over there.

And sure, if you feel such a concrete Kendall here does not make so much sense, we can move it; we are working towards a breaking release here anyways.

kellertuer commented:
My problem with moving Grassmann is that it really would break a lot a lot of code, including my current PR to DFTK.

That is why I proposed to have a simple or the current version just here and a boosted/enhanced version in Liegroups.

I mean the current Grassmann version here works quite okay and fine – just without LieGroups. I do not see a reason to remove it, just because in principle it is a quotient. Same for the sphere. And maybe even for Kendall. But Kendall seems to actually need LieGroups tools, the other two do not directly need that.

kellertuer committed:
Merge branch 'master' into kellertuer/runic

# Conflicts:
#	.github/workflows/format.yml

kellertuer committed:
Merge branch 'kellertuer/runic' into mbaran/trait-reduction

# Conflicts:
#	NEWS.md
#	ext/ManifoldsOrdinaryDiffEqDiffEqCallbacksExt.jl
#	ext/ManifoldsOrdinaryDiffEqExt.jl
#	ext/ManifoldsRecursiveArrayToolsExt/ProductGroupRAT.jl
#	ext/ManifoldsRecursiveArrayToolsExt/rotation_translation_action_rat.jl
#	ext/ManifoldsRecursiveArrayToolsExt/special_euclidean_rat.jl
#	ext/ManifoldsTestExt/tests_general.jl
#	ext/ManifoldsTestExt/tests_group.jl
#	src/groups/GroupManifold.jl
#	src/groups/addition_operation.jl
#	src/groups/circle_group.jl
#	src/groups/connections.jl
#	src/groups/general_linear.jl
#	src/groups/general_unitary_groups.jl
#	src/groups/group.jl
#	src/groups/group_action.jl
#	src/groups/group_operation_action.jl
#	src/groups/heisenberg.jl
#	src/groups/metric.jl
#	src/groups/multiplication_operation.jl
#	src/groups/orthogonal.jl
#	src/groups/power_group.jl
#	src/groups/product_group.jl
#	src/groups/rotation_action.jl
#	src/groups/rotation_translation_action.jl
#	src/groups/semidirect_product_group.jl
#	src/groups/special_euclidean.jl
#	src/groups/special_linear.jl
#	src/groups/special_orthogonal.jl
#	src/groups/special_unitary.jl
#	src/groups/translation_action.jl
#	src/groups/translation_group.jl
#	src/groups/unitary.jl
#	src/groups/validation_group.jl
#	src/manifolds/ConnectionManifold.jl
#	src/manifolds/Elliptope.jl
#	src/manifolds/GeneralUnitaryMatrices.jl
#	src/manifolds/Grassmann.jl
#	src/manifolds/GrassmannProjector.jl
#	src/manifolds/HyperbolicPoincareBall.jl
#	src/manifolds/HyperbolicPoincareHalfspace.jl
#	src/manifolds/KendallsShapeSpace.jl
#	src/manifolds/MetricManifold.jl
#	src/manifolds/ProductManifold.jl
#	src/manifolds/QuotientManifold.jl
#	src/manifolds/Stiefel.jl
#	src/manifolds/Symmetric.jl
#	src/manifolds/Symplectic.jl
#	src/statistics.jl
#	src/trait_recursion_breaking.jl
#	test/groups/circle_group.jl
#	test/groups/connections.jl
#	test/groups/general_linear.jl
#	test/groups/general_unitary_groups.jl
#	test/groups/group_operation_action.jl
#	test/groups/group_utils.jl
#	test/groups/groups_general.jl
#	test/groups/heisenberg.jl
#	test/groups/metric.jl
#	test/groups/power_group.jl
#	test/groups/product_group.jl
#	test/groups/rotation_action.jl
#	test/groups/rotation_translation_action.jl
#	test/groups/special_euclidean.jl
#	test/groups/special_linear.jl
#	test/groups/special_orthogonal.jl
#	test/groups/translation_action.jl
#	test/groups/translation_group.jl
#	test/groups/validation_group.jl
#	test/manifolds/grassmann.jl
#	test/manifolds/shape_space.jl
#	test/manifolds/skewhermitian.jl
#	test/manifolds/symmetric_positive_semidefinite_fixed_rank.jl
#	test/metric.jl

kellertuer committed:
run formatter on the new code as well.

kellertuer committed:
Merge branch 'master' into mbaran/trait-reduction

# Conflicts:
#	NEWS.md
#	ext/ManifoldsOrdinaryDiffEqDiffEqCallbacksExt.jl
#	ext/ManifoldsOrdinaryDiffEqExt.jl
#	ext/ManifoldsRecursiveArrayToolsExt/ProductGroupRAT.jl
#	ext/ManifoldsRecursiveArrayToolsExt/rotation_translation_action_rat.jl
#	ext/ManifoldsRecursiveArrayToolsExt/special_euclidean_rat.jl
#	ext/ManifoldsTestExt/tests_general.jl
#	ext/ManifoldsTestExt/tests_group.jl
#	src/groups/GroupManifold.jl
#	src/groups/addition_operation.jl
#	src/groups/circle_group.jl
#	src/groups/connections.jl
#	src/groups/general_linear.jl
#	src/groups/general_unitary_groups.jl
#	src/groups/group.jl
#	src/groups/group_action.jl
#	src/groups/group_operation_action.jl
#	src/groups/heisenberg.jl
#	src/groups/metric.jl
#	src/groups/multiplication_operation.jl
#	src/groups/orthogonal.jl
#	src/groups/power_group.jl
#	src/groups/product_group.jl
#	src/groups/rotation_action.jl
#	src/groups/rotation_translation_action.jl
#	src/groups/semidirect_product_group.jl
#	src/groups/special_euclidean.jl
#	src/groups/special_linear.jl
#	src/groups/special_orthogonal.jl
#	src/groups/special_unitary.jl
#	src/groups/translation_action.jl
#	src/groups/translation_group.jl
#	src/groups/unitary.jl
#	src/groups/validation_group.jl
#	src/manifolds/CenteredMatrices.jl
#	src/manifolds/ConnectionManifold.jl
#	src/manifolds/DeterminantOneMatrices.jl
#	src/manifolds/Elliptope.jl
#	src/manifolds/Euclidean.jl
#	src/manifolds/FixedRankMatrices.jl
#	src/manifolds/Flag.jl
#	src/manifolds/FlagOrthogonal.jl
#	src/manifolds/GeneralUnitaryMatrices.jl
#	src/manifolds/GeneralizedGrassmann.jl
#	src/manifolds/GeneralizedStiefel.jl
#	src/manifolds/Grassmann.jl
#	src/manifolds/GrassmannProjector.jl
#	src/manifolds/GrassmannStiefel.jl
#	src/manifolds/Hamiltonian.jl
#	src/manifolds/HeisenbergMatrices.jl
#	src/manifolds/Hyperbolic.jl
#	src/manifolds/HyperbolicPoincareBall.jl
#	src/manifolds/HyperbolicPoincareHalfspace.jl
#	src/manifolds/InvertibleMatrices.jl
#	src/manifolds/KendallsPreShapeSpace.jl
#	src/manifolds/KendallsShapeSpace.jl
#	src/manifolds/MetricManifold.jl
#	src/manifolds/MultinomialDoublyStochastic.jl
#	src/manifolds/MultinomialSymmetric.jl
#	src/manifolds/MultinomialSymmetricPositiveDefinite.jl
#	src/manifolds/ProbabilitySimplex.jl
#	src/manifolds/ProductManifold.jl
#	src/manifolds/ProjectiveSpace.jl
#	src/manifolds/QuotientManifold.jl
#	src/manifolds/SPDFixedDeterminant.jl
#	src/manifolds/SkewHermitian.jl
#	src/manifolds/Spectrahedron.jl
#	src/manifolds/Sphere.jl
#	src/manifolds/SphereSymmetricMatrices.jl
#	src/manifolds/Stiefel.jl
#	src/manifolds/Symmetric.jl
#	src/manifolds/SymmetricPositiveDefinite.jl
#	src/manifolds/SymmetricPositiveSemidefiniteFixedRank.jl
#	src/manifolds/Symplectic.jl
#	src/manifolds/SymplecticGrassmann.jl
#	src/manifolds/SymplecticGrassmannProjector.jl
#	src/manifolds/SymplecticGrassmannStiefel.jl
#	src/manifolds/SymplecticStiefel.jl
#	src/statistics.jl
#	src/trait_recursion_breaking.jl
#	test/groups/circle_group.jl
#	test/groups/connections.jl
#	test/groups/general_linear.jl
#	test/groups/general_unitary_groups.jl
#	test/groups/group_operation_action.jl
#	test/groups/group_utils.jl
#	test/groups/groups_general.jl
#	test/groups/heisenberg.jl
#	test/groups/metric.jl
#	test/groups/power_group.jl
#	test/groups/product_group.jl
#	test/groups/rotation_action.jl
#	test/groups/rotation_translation_action.jl
#	test/groups/special_euclidean.jl
#	test/groups/special_linear.jl
#	test/groups/special_orthogonal.jl
#	test/groups/translation_action.jl
#	test/groups/translation_group.jl
#	test/groups/validation_group.jl
#	test/manifolds/grassmann.jl
#	test/manifolds/shape_space.jl
#	test/manifolds/skewhermitian.jl
#	test/manifolds/symmetric_positive_semidefinite_fixed_rank.jl
#	test/metric.jl

kellertuer commented:
Merged runic, brought it here.

For me locally (with the other ManifoldsBase branch checked out) some small things on Grassmann and Kendall still fail (one of them due to the orbit-function that was deleted, Grassmann with exp?) but maybe that was there before. Can look at it closer tomorrow.

kellertuer committed:
Fix export of inv, inv!, since they are still used for the symplectic inverse.

kellertuer committed:
Rename new_trait_function to trait_function.

mateuszbaran committed:
adapt to renaming in ManifoldsBase

kellertuer committed:
Bump version and dependencies, add more details to the news what was moved and is breaking.

kellertuer committed:
fix imports.

kellertuer committed:
Remove two obsolete tests.

kellertuer committed:
Remove totalspace, adapt a few expected errors.

kellertuer committed:
runic.

kellertuer committed:
Fix a few more tests.

kellertuer committed:
Remove deprecated types and mention that in the news (again).

kellertuer commented:
With ManifoldsBase 2.0 this PR only errors on the distance of Kendalls shape space, since we do not longer have the optimal alignment available here. Can we maybe hardcode this specific case in that distance to make it work again?

All other tests run locally for me (but I also have to switch parameters).

kellertuer committed:
remove obsolete group `test_lie_groups` from CI and nightly.

kellertuer committed:
Remove quoeitnet manifold type and the function definitions that have been moved to ManifoldsBase.

kellertuer committed:
remove an allocate I missed.

mateuszbaran commented:
IIRC `log` also won't work without `optimal_alignment`. I considered doing it locally here but I don't like it because `optimal_alignment` is the natural place to define that functionality and we'd duplicate some code or do weird dispatches. I don't like other solutions either so I'm not strongly opposed to this one. Maybe we could have `LieGroupsBase.jl` that defines `optimal_alignment` and is a dependency of `Manifolds.jl`? That's more work but otherwise it might be the cleanest  solution.

kellertuer commented:
Hm. We are Tendenz towards more and more Detroits. Can we finish 11 where we delete these and do that as a next step?

mateuszbaran commented:
That's fine too, sure.

kellertuer committed:
Start rearranging parameters until E – Euclidean.

kellertuer committed:
Fix all tests that checked for the concrete parameter order in Eucldiean.

kellertuer committed:
Fix a few more type tests.

kellertuer committed:
Fix Kendalls shape space to (at least for now) work without the whole Lie groups action machinery.

kellertuer committed:
Fix diff canonical project on Grassmann.

kellertuer committed:
Fix two more tests.

kellertuer committed:
Fix one further test.

mateuszbaran committed:
fix a test on a power manifold

kellertuer committed:
Continue refactor of type parameters until and including G - and manifolds based on those like Stiefel or rotations.

kellertuer committed:
run runic.

kellertuer committed:
Fix distributions and a grassmann test.

kellertuer committed:
Fix a few places where I was not careful enough in refactoring.

kellertuer committed:
Finish all manifolds until and including the Multinomial ones to have a consistent parameter order.

kellertuer committed:
Fix all manifolds up to and including Segre; and the Sphere, since is used within Oblique.

kellertuer committed:
finish remaining manifolds.

kellertuer commented:
I finished the refactor of all manifolds to have the F upfront. I noticed that I either had to switch or could simplify some `ManifoldName{<:Any, F}` occurrences.
All those changes are documented.

So now locally only a few statistics methods fail. that I do not fully understand why they do.

I think we can think about releasing ManifoldsBase 2.0, announcing that and looking here for code coverage (and bumping the remaining packages).

kellertuer committed:
already bump dependencies for Manifolds and ManifoldsBase in docs

kellertuer committed:
Fix some more tests

mateuszbaran commented on ext/ManifoldsOrdinaryDiffEqExt.jl:
```diff
@@ -45,13 +44,25 @@ function solve_exp_ode(
-        ::TraitList{IsMetricManifold}, M::AbstractDecoratorManifold, q, p, X, t::Number;
-        kwargs...
+
+function ManifoldsBase.retract_exp_ode!(
```
Since we're making a breaking release here, I think we could finally remove this solver? `solve_chart_exp_ode` should be the replacement.

mateuszbaran committed:
remove is_embedded method

mateuszbaran committed:
that method isn't needed

mateuszbaran committed:
fix statistics tests

kellertuer committed:
Fix oblique, sphere symmetric, and tucker.

kellertuer committed:
Merge branch 'mbaran/trait-reduction' of github.com:JuliaManifolds/Manifolds.jl into mbaran/trait-reduction

mateuszbaran committed:
remove ODEExponentialRetraction in favor of solving in charts

kellertuer committed:
Merge branch 'mbaran/trait-reduction' of github.com:JuliaManifolds/Manifolds.jl into mbaran/trait-reduction

kellertuer committed:
Set release date for 0.11.0

kellertuer committed:
Fix reciepts.

mateuszbaran committed:
fix recipes fix

mateuszbaran committed:
update tutorials compat

kellertuer committed:
Merge branch 'master' into mbaran/trait-reduction

kellertuer committed:
Also remove architecture from nightly.

kellertuer committed:
Remove the GroupManifolds tutorial.

kellertuer committed:
bump project.

kellertuer committed:
Remove further groupmanifold artefacts from the docs.

kellertuer committed:
Fix some more docs.

kellertuer committed:
Remove more artefacts from former group manifolds

mateuszbaran committed:
fix metric manifold tests

mateuszbaran committed:
remove `Identity` ambiguity filter

mateuszbaran committed:
more detailed report on check_vector failure

kellertuer committed:
Fix docs warnings.

kellertuer committed:
remove warnonly for missing docs, since all are included now.

kellertuer committed:
Switch to QuaertoNotebookRunner.

kellertuer committed:
Runs runic.

mateuszbaran commented on docs/make.jl:
```diff
@@ -110,6 +101,22 @@ using Test, FiniteDifferences
 
 # (e) add CONTRIBUTING.md and NEWS.md to docs
+
+function add_links(line::String, url::String = "https://github.com/JuliaManifolds/Manopt.jl")
```
Shouldn't it link to Manifolds.jl instead of Manopt.jl?

kellertuer committed:
Fix a link url

kellertuer committed:
Fix sources.

kellertuer committed:
Stop caching Conda.

kellertuer committed:
Also remove CondaPkg here.

kellertuer committed:
put defaultmetric into the right spot.

kellertuer committed:
Bump quarto version for documenter.

kellertuer committed:
Just some very few lines code cov.

mateuszbaran commented on test/groups/general_unitary_groups.jl:
```diff
@@ -1,261 +0,0 @@
-        @test manifold_volume(SpecialUnitary(4)) ≈ sqrt(2) * 4 * π^9 / 12
-    end
-
-    @testset "SO(4) and O(4) exp/log edge cases" begin
```
Should we keep these tests in Manifolds.jl as tests for `exp`/`log` or move them to LieGroups.jl?

kellertuer replied:
I think they are already done as edge cases on Liegroups.jl? See e.g. https://github.com/JuliaManifolds/LieGroups.jl/blob/main/test/LieGroupsTestSuite.jl/so4_edge_cases.jl where they are defined even within the test suite.

kellertuer replied:
Maybe you are right and added them back already for the matrices. So we can close this.

mateuszbaran committed:
volume and volume density tests; SpecialUnitaryMatrices

mateuszbaran commented on docs/src/manifolds/shapespace.md:
```diff
@@ -21,12 +21,6 @@ q = [
 fig = scatter(p[1,:], p[2,:], label="p", aspect_ratio=:equal)
 scatter!(fig, q[1,:], q[2,:], label="q")
-
-# aligning q to p
-A = get_orbit_action(M)
-a = optimal_alignment(A, p, q)
-rot_q = apply(A, a, q)
-scatter!(fig, rot_q[1,:], rot_q[2,:], label="q aligned to p")
```
This is a relatively important part of shape spaces, so maybe it would be better to move the entire tutorial to LieGroups.jl instead of keeping an incomplete one here.

kellertuer replied:
Ah yes, sorry, I did not have that in mind, that it might make the example too minimalistic.

I think it would be great example for ManifoldExamples – maybe even defining the optimal alignment just in the example might work for now.

IN general I think, having more like “how to do this” tutorials here and nice examples over on the examples packages might be a good idea to follow.

kellertuer replied:
I understand and sure then lets move it over there.

kellertuer replied:
Ok, so then let's just remove code of the example here.

kellertuer committed:
Cover some lines of code.

kellertuer committed:
Fix docs.

kellertuer committed:
Work on a few lines of test coverage.

kellertuer committed:
A bit of code cov a day keeps the bug away.

kellertuer committed:
continued testing.

kellertuer committed:
a few more lines

kellertuer commented:
The last 23 lines are a bit hard to spot since code cov is a bit confused and there is a few more lines than that not covered. Do you want to give it a try? I want at least through easy cases I could check.

mateuszbaran committed:
add some tests, organize code a bit better

mateuszbaran commented:
I've added a few tests. It's easier to see what's missing by looking at the entire file on codecov like here: https://app.codecov.io/gh/JuliaManifolds/Manifolds.jl/pull/826/blob/src/manifolds/GeneralUnitaryMatrices.jl instead of the default overview.

kellertuer committed:
a few more tests.

kellertuer committed:
unify mul! style.

kellertuer committed:
Fix a typo, add more tests.

kellertuer committed:
codecov for spds.

mateuszbaran committed:
more tests

mateuszbaran committed:
Merge remote-tracking branch 'origin/mbaran/trait-reduction' into mbaran/trait-reduction

kellertuer committed:
Adress points from review.

kellertuer committed:
Merge branch 'mbaran/trait-reduction' of github.com:JuliaManifolds/Manifolds.jl into mbaran/trait-reduction
